src/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Net.Http.Headers
+{
+    // This struct represents a particular named header --
+    // if the header is one of our known headers, then it contains a reference to the KnownHeader object;
+    // otherwise, for custom headers, it just contains a string for the header name.
+    // Use HeaderDescriptor.Get to resolve an arbitrary header name to a HeaderDescriptor.
+    internal struct HeaderDescriptor : IEquatable<HeaderDescriptor>
+    {
+        private readonly string _headerName;
+        private readonly KnownHeader _knownHeader;
+
+        public HeaderDescriptor(KnownHeader knownHeader)
+        {
+            _knownHeader = knownHeader;
+            _headerName = knownHeader.Name;
+        }
+
+        // This should not be used directly; use static Get below
+        private HeaderDescriptor(string headerName)
+        {
+            _headerName = headerName;
+            _knownHeader = null;
+        }
+
+        public string Name => _headerName;
+        public HttpHeaderParser Parser => _knownHeader?.Parser;
+        public HttpHeaderType HeaderType => _knownHeader == null ? HttpHeaderType.Custom : _knownHeader.HeaderType;
+
+        public bool Equals(HeaderDescriptor other) =>
+            (_knownHeader == null ?
+                StringComparer.OrdinalIgnoreCase.Equals(_headerName, other._headerName) :
+                _knownHeader == other._knownHeader);
+        public override int GetHashCode() => (_knownHeader == null ? StringComparer.OrdinalIgnoreCase.GetHashCode(_headerName) : _knownHeader.GetHashCode());
+        public override bool Equals(object obj) => (obj is HeaderDescriptor headerInfo ? Equals(headerInfo) : false);
+        public static bool operator ==(HeaderDescriptor left, HeaderDescriptor right) => left.Equals(right);
+        public static bool operator !=(HeaderDescriptor left, HeaderDescriptor right) => !left.Equals(right);
+
+        public static bool TryGet(string name, out HeaderDescriptor descriptor)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                descriptor = default(HeaderDescriptor);
+                return false;
+            }
+
+            if (HttpRuleParser.GetTokenLength(name, 0) != name.Length)
+            {
+                descriptor = default(HeaderDescriptor);
+                return false;
+            }
+
+            KnownHeader knownHeader = KnownHeaders.TryGetKnownHeader(name);
+            if (knownHeader != null)
+            {
+                descriptor = new HeaderDescriptor(knownHeader);
```

Comment:
With HeaderDescriptor containing several references, this might end up incurring some write barriers.  May not amount to anything measurable, but it's top of mind for me as I did find a measurable case of this in some other changes recently.